kclvm-runtime fails to buid on aarch64

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

On an aarch64 machine (I used a t4g.medium AWS EC2 instance):

```sh
git clone --depth=1 https://github.com/kcl-lang/kcl.git  && cd kcl   && make build
```

### 2. What did you expect to see? (Required)

Compilation succeeds as on x86_64

### 3. What did you see instead (Required)

```
  Compiling kclvm-runtime v0.7.3 (/root/kcl/kclvm/runtime)
error[E0308]: mismatched types
  --> runtime/src/api/utils.rs:39:47
   |
39 |     let s = unsafe { std::ffi::CStr::from_ptr(s) }.to_str().unwrap();
   |                      ------------------------ ^ expected `*const u8`, found `*const i8`
   |                      |
   |                      arguments to this function are incorrect
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`
note: associated function defined here
  --> /build/rustc-ZBSOmY/rustc-1.70.0+dfsg0ubuntu1~bpo2/library/core/src/ffi/c_str.rs:260:25

error[E0308]: mismatched types
   --> runtime/src/context/api.rs:195:24
    |
195 |     let method = c2str(method);
    |                  ----- ^^^^^^ expected `*const i8`, found `*const u8`
    |                  |
    |                  arguments to this function are incorrect
    |
    = note: expected raw pointer `*const i8`
               found raw pointer `*const u8`
note: function defined here
   --> runtime/src/api/utils.rs:38:15
    |
38  | pub(crate) fn c2str<'a>(s: *const i8) -> &'a str {
    |               ^^^^^     ------------

error[E0308]: mismatched types
   --> runtime/src/context/api.rs:197:41
    |
197 |     let args = kclvm_value_from_json(p, args);
    |                ---------------------    ^^^^ expected `*const i8`, found `*const u8`
    |                |
    |                arguments to this function are incorrect
    |
    = note: expected raw pointer `*const i8`
               found raw pointer `*const u8`
note: function defined here
   --> runtime/src/value/api.rs:432:26
    |
432 | pub unsafe extern "C" fn kclvm_value_from_json(
    |                          ^^^^^^^^^^^^^^^^^^^^^
433 |     ctx: *mut kclvm_context_t,
434 |     s: *const kclvm_char_t,
    |     ----------------------

error[E0308]: mismatched types
   --> runtime/src/context/api.rs:198:43
    |
198 |     let kwargs = kclvm_value_from_json(p, kwargs);
    |                  ---------------------    ^^^^^^ expected `*const i8`, found `*const u8`
    |                  |
    |                  arguments to this function are incorrect
    |
    = note: expected raw pointer `*const i8`
               found raw pointer `*const u8`
note: function defined here
   --> runtime/src/value/api.rs:432:26
    |
432 | pub unsafe extern "C" fn kclvm_value_from_json(
    |                          ^^^^^^^^^^^^^^^^^^^^^
433 |     ctx: *mut kclvm_context_t,
434 |     s: *const kclvm_char_t,
    |     ----------------------

error[E0308]: mismatched types
   --> runtime/src/context/api.rs:208:5
    |
193 | ) -> *const c_char {
    |      ------------- expected `*const u8` because of return type
...
208 |     result_json
    |     ^^^^^^^^^^^ expected `*const u8`, found `*const i8`
    |
    = note: expected raw pointer `*const u8`
               found raw pointer `*const i8`

error[E0308]: arguments to this function are incorrect
  --> runtime/src/stdlib/plugin.rs:44:23
   |
44 |     let result_json = kclvm_plugin_invoke_json(method, args_json, kwargs_json);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: expected `*const u8`, found `*const i8`
  --> runtime/src/stdlib/plugin.rs:44:56
   |
44 |     let result_json = kclvm_plugin_invoke_json(method, args_json, kwargs_json);
   |                                                        ^^^^^^^^^
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`
note: expected `*const u8`, found `*const i8`
  --> runtime/src/stdlib/plugin.rs:44:67
   |
44 |     let result_json = kclvm_plugin_invoke_json(method, args_json, kwargs_json);
   |                                                                   ^^^^^^^^^^^
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`
note: function defined here
  --> runtime/src/stdlib/plugin.rs:66:19
   |
66 | pub extern "C" fn kclvm_plugin_invoke_json(
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^
67 |     method: *const i8,
   |     -----------------
68 |     args: *const c_char,
   |     -------------------
69 |     kwargs: *const c_char,
   |     ---------------------

error[E0308]: mismatched types
   --> runtime/src/stdlib/plugin.rs:50:42
    |
50  |     let ptr = kclvm_value_from_json(ctx, result_json);
    |               ---------------------      ^^^^^^^^^^^ expected `*const i8`, found `*const u8`
    |               |
    |               arguments to this function are incorrect
    |
    = note: expected raw pointer `*const i8`
               found raw pointer `*const u8`
note: function defined here
   --> runtime/src/value/api.rs:432:26
    |
432 | pub unsafe extern "C" fn kclvm_value_from_json(
    |                          ^^^^^^^^^^^^^^^^^^^^^
433 |     ctx: *mut kclvm_context_t,
434 |     s: *const kclvm_char_t,
    |     ----------------------

For more information about this error, try `rustc --explain E0308`.
error: could not compile `kclvm-runtime` (lib) due to 7 previous errors
warning: build failed, waiting for other jobs to finish...
make: *** [Makefile:26: build] Error 101
subprocess exited with status 2
subprocess exited with status 2
```

### 4. What is your KCL components version? (Required)

kcl.git 174204777a1d29d7dfeb74d4f88f08a962cea0e0

Thanks for the quick response!  With the latest commit, I'm seeing some different but related compilation errors:

```
error[E0308]: mismatched types
  --> runtime/src/context/api.rs:93:50
   |
93 |         p.set_kcl_location_info(None, Some(c2str(filename)), Some(line), Some(col));
   |                                            ----- ^^^^^^^^ expected `*const u8`, found `*const i8`
   |                                            |
   |                                            arguments to this function are incorrect
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`
note: function defined here
  --> runtime/src/api/utils.rs:40:15
   |
40 | pub(crate) fn c2str<'a>(s: *const c_char) -> &'a str {
   |               ^^^^^     ----------------

error[E0308]: mismatched types
   --> runtime/src/context/api.rs:107:33
    |
107 |         p.set_kcl_pkgpath(c2str(pkgpath));
    |                           ----- ^^^^^^^ expected `*const u8`, found `*const i8`
    |                           |
    |                           arguments to this function are incorrect
    |
    = note: expected raw pointer `*const u8`
               found raw pointer `*const i8`
note: function defined here
   --> runtime/src/api/utils.rs:40:15
    |
40  | pub(crate) fn c2str<'a>(s: *const c_char) -> &'a str {
    |               ^^^^^     ----------------

error[E0308]: mismatched types
   --> runtime/src/context/api.rs:119:36
    |
119 |         ctx.set_kcl_filename(c2str(filename));
    |                              ----- ^^^^^^^^ expected `*const u8`, found `*const i8`
    |                              |
    |                              arguments to this function are incorrect
    |
    = note: expected raw pointer `*const u8`
               found raw pointer `*const i8`
note: function defined here
   --> runtime/src/api/utils.rs:40:15
    |
40  | pub(crate) fn c2str<'a>(s: *const c_char) -> &'a str {
    |               ^^^^^     ----------------

error[E0308]: mismatched types
   --> runtime/src/context/api.rs:208:5
    |
193 | ) -> *const c_char {
    |      ------------- expected `*const u8` because of return type
...
208 |     result_json
    |     ^^^^^^^^^^^ expected `*const u8`, found `*const i8`
    |
    = note: expected raw pointer `*const u8`
               found raw pointer `*const i8`

error[E0308]: mismatched types
   --> runtime/src/value/api.rs:196:33
    |
196 |         if v.is_null() || *v == '\0' as i8 {
    |                                 ^^^^^^^^^^ expected `u8`, found `i8`

error[E0277]: can't compare `u8` with `i8`
   --> runtime/src/value/api.rs:196:30
    |
196 |         if v.is_null() || *v == '\0' as i8 {
    |                              ^^ no implementation for `u8 == i8`
    |
    = help: the trait `PartialEq<i8>` is not implemented for `u8`
    = help: the following other types implement trait `PartialEq<Rhs>`:
              <u8 as PartialEq<serde_json::Value>>
              <u8 as PartialEq>

error[E0308]: mismatched types
   --> runtime/src/value/api.rs:546:36
    |
543 | pub unsafe extern "C" fn kclvm_value_Str_ptr(p: *const kclvm_value_ref_t) -> *const kclvm_char_t {
    |                                                                              ------------------- expected `*const u8` because of return type
...
546 |         Value::str_value(ref v) => v.as_ptr() as *const i8,
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
    |
    = note: expected raw pointer `*const u8`
               found raw pointer `*const i8`

error[E0308]: mismatched types
  --> runtime/src/stdlib/builtin_api.rs:17:35
   |
17 |     ctx.builtin_option_init(c2str(key), c2str(value));
   |                             ----- ^^^ expected `*const u8`, found `*const i8`
   |                             |
   |                             arguments to this function are incorrect
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`
note: function defined here
  --> runtime/src/api/utils.rs:40:15
   |
40 | pub(crate) fn c2str<'a>(s: *const c_char) -> &'a str {
   |               ^^^^^     ----------------

error[E0308]: mismatched types
  --> runtime/src/stdlib/builtin_api.rs:17:47
   |
17 |     ctx.builtin_option_init(c2str(key), c2str(value));
   |                                         ----- ^^^^^ expected `*const u8`, found `*const i8`
   |                                         |
   |                                         arguments to this function are incorrect
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`
note: function defined here
  --> runtime/src/api/utils.rs:40:15
   |
40 | pub(crate) fn c2str<'a>(s: *const c_char) -> &'a str {
   |               ^^^^^     ----------------

error[E0308]: mismatched types
  --> runtime/src/stdlib/plugin.rs:44:48
   |
44 |     let result_json = kclvm_plugin_invoke_json(method, args_json, kwargs_json);
   |                       ------------------------ ^^^^^^ expected `*const u8`, found `*const i8`
   |                       |
   |                       arguments to this function are incorrect
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`
note: function defined here
  --> runtime/src/stdlib/plugin.rs:66:19
   |
66 | pub extern "C" fn kclvm_plugin_invoke_json(
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^
67 |     method: *const c_char,
   |     ---------------------

error[E0308]: mismatched types
   --> runtime/src/_kcl_run.rs:190:34
    |
190 |             kclvm_value_Str(ctx, b"{}\0" as *const u8 as *const kclvm_char_t)
    |             ---------------      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
    |             |
    |             arguments to this function are incorrect
    |
    = note: expected raw pointer `*const u8`
               found raw pointer `*const i8`
note: function defined here
   --> runtime/src/value/api.rs:190:26
    |
190 | pub unsafe extern "C" fn kclvm_value_Str(
    |                          ^^^^^^^^^^^^^^^
191 |     ctx: *mut kclvm_context_t,
192 |     v: *const kclvm_char_t,
    |     ----------------------

error[E0308]: mismatched types
   --> runtime/src/_kcl_run.rs:203:39
    |
203 |             std::ptr::copy(c_str_ptr, result_buffer, return_len as usize);
    |             --------------            ^^^^^^^^^^^^^ expected `*mut u8`, found `*mut i8`
    |             |
    |             arguments to this function are incorrect
    |
    = note: expected raw pointer `*mut u8`
               found raw pointer `*mut i8`
note: function defined here
   --> /build/rustc-ZBSOmY/rustc-1.70.0+dfsg0ubuntu1~bpo2/library/core/src/intrinsics.rs:2710:21
```

--- a/kclvm/api/src/service/capi.rs
+++ b/kclvm/api/src/service/capi.rs
@@ -90,7 +90,6 @@ pub extern "C" fn kclvm_service_call(
         unsafe { (*call)(serv, args) }
     });
     match result {
-        //todo uniform error handling
         Ok(result) => result,
         Err(panic_err) => {
             let err_message = kclvm_error::err_to_str(panic_err);
@@ -98,7 +97,7 @@ pub extern "C" fn kclvm_service_call(
             let c_string = std::ffi::CString::new(format!("ERROR:{}", err_message.as_str()))
                 .expect("CString::new failed");
             let ptr = c_string.into_raw();
-            ptr as *const i8
+            ptr as *const c_char
         }
     }
 }

--- a/kclvm/runner/src/runner.rs
+++ b/kclvm/runner/src/runner.rs
@@ -11,11 +11,12 @@ use kclvm_query::r#override::parse_override_spec;
 use kclvm_runtime::{Context, PanicInfo, ValueRef};
 use serde::{Deserialize, Serialize};
 use std::ffi::OsStr;
+use std::os::raw::c_char;
 
 const RESULT_SIZE: usize = 2048 * 2048;
 
 #[allow(non_camel_case_types)]
-pub type kclvm_char_t = i8;
+pub type kclvm_char_t = c_char;
 #[allow(non_camel_case_types)]
 pub type kclvm_size_t = i32;
 #[allow(non_camel_case_types)]
@@ -357,17 +358,17 @@ impl KclLibRunner {
         // Exec json result
         let mut json_result = vec![0u8; RESULT_SIZE];
         let mut result_buffer_len = json_result.len() as i32 - 1;
-        let json_result_buffer = json_result.as_mut_ptr() as *mut i8;
+        let json_result_buffer = json_result.as_mut_ptr() as *mut c_char;
 
         // Exec warning data
         let mut warn_data = vec![0u8; RESULT_SIZE];
         let mut warn_buffer_len = warn_data.len() as i32 - 1;
-        let warn_buffer = warn_data.as_mut_ptr() as *mut i8;
+        let warn_buffer = warn_data.as_mut_ptr() as *mut c_char;
 
         // Exec log data
         let mut log_data = vec![0u8; RESULT_SIZE];
         let mut log_buffer_len = log_data.len() as i32 - 1;
-        let log_buffer = log_data.as_mut_ptr() as *mut i8;
+        let log_buffer = log_data.as_mut_ptr() as *mut c_char;
 
         let n = kcl_run(
             kclvm_main_ptr,

--- a/kclvm/runtime/src/_kcl_run.rs
+++ b/kclvm/runtime/src/_kcl_run.rs
@@ -1,6 +1,8 @@
 // Copyright 2021 The KCL Authors. All rights reserved.
 #![allow(clippy::missing_safety_doc)]
 
+use std::os::raw::c_char;
+
 use crate::*;
 
 #[allow(dead_code, non_camel_case_types)]
@@ -19,7 +21,7 @@ type kclvm_value_ref_t = ValueRef;
 type kclvm_iterator_t = ValueIterator;
 
 #[allow(dead_code, non_camel_case_types)]
-type kclvm_char_t = i8;
+type kclvm_char_t = c_char;
 
 #[allow(dead_code, non_camel_case_types)]
 type kclvm_size_t = i32;
@@ -115,7 +117,7 @@ pub unsafe extern "C" fn _kcl_run(
     // Get the runtime context.
     let ctx_ref = ptr_as_ref(ctx);
     // Copy log message pointer
-    let c_str_ptr = ctx_ref.log_message.as_ptr() as *const i8;
+    let c_str_ptr = ctx_ref.log_message.as_ptr() as *const c_char;
     let c_str_len = ctx_ref.log_message.len() as i32;
     if c_str_len <= *log_buffer_len {
         std::ptr::copy(c_str_ptr, log_buffer, c_str_len as usize);
@@ -123,7 +125,7 @@ pub unsafe extern "C" fn _kcl_run(
     }
     // Copy panic info message pointer
     let json_panic_info = ctx_ref.get_panic_info_json_string();
-    let c_str_ptr = json_panic_info.as_ptr() as *const i8;
+    let c_str_ptr = json_panic_info.as_ptr() as *const c_char;
     let c_str_len = json_panic_info.len() as i32;
     match result {
         Ok(n) => {

--- a/kclvm/runtime/src/context/api.rs
+++ b/kclvm/runtime/src/context/api.rs
@@ -84,7 +84,7 @@ pub unsafe extern "C" fn kclvm_context_main_end_hook(
 #[runtime_fn]
 pub unsafe extern "C" fn kclvm_context_set_kcl_location(
     p: *mut kclvm_context_t,
-    filename: *const i8,
+    filename: *const c_char,
     line: i32,
     col: i32,
 ) {
@@ -100,7 +100,7 @@ pub unsafe extern "C" fn kclvm_context_set_kcl_location(
 #[runtime_fn]
 pub unsafe extern "C" fn kclvm_context_set_kcl_pkgpath(
     p: *mut kclvm_context_t,
-    pkgpath: *const i8,
+    pkgpath: *const c_char,
 ) {
     let p = mut_ptr_as_ref(p);
     if !pkgpath.is_null() {
@@ -112,7 +112,7 @@ pub unsafe extern "C" fn kclvm_context_set_kcl_pkgpath(
 #[runtime_fn]
 pub unsafe extern "C" fn kclvm_context_set_kcl_filename(
     ctx: *mut kclvm_context_t,
-    filename: *const i8,
+    filename: *const c_char,
 ) {
     let ctx = mut_ptr_as_ref(ctx);
     if !filename.is_null() {
@@ -199,7 +199,7 @@ pub unsafe extern "C" fn kclvm_context_invoke(
     let result = _kclvm_context_invoke(p, method, args, kwargs);
 
     p.buffer.kclvm_context_invoke_result = ptr_as_ref(result).to_json_string_with_null();
-    let result_json = p.buffer.kclvm_context_invoke_result.as_ptr() as *const i8;
+    let result_json = p.buffer.kclvm_context_invoke_result.as_ptr() as *const c_char;
 
     kclvm_value_delete(args);
     kclvm_value_delete(kwargs);

--- a/kclvm/runtime/src/stdlib/builtin_api.rs
+++ b/kclvm/runtime/src/stdlib/builtin_api.rs
@@ -1,6 +1,8 @@
 // Copyright 2021 The KCL Authors. All rights reserved.
 #![allow(clippy::missing_safety_doc)]
 
+use std::os::raw::c_char;
+
 use crate::*;
 
 #[allow(non_camel_case_types)]
@@ -10,8 +12,8 @@ type kclvm_value_ref_t = ValueRef;
 #[runtime_fn]
 pub unsafe extern "C" fn kclvm_builtin_option_init(
     ctx: *mut kclvm_context_t,
-    key: *const i8,
-    value: *const i8,
+    key: *const c_char,
+    value: *const c_char,
 ) {
     let ctx = mut_ptr_as_ref(ctx);
     ctx.builtin_option_init(c2str(key), c2str(value));

--- a/kclvm/runtime/src/stdlib/plugin.rs
+++ b/kclvm/runtime/src/stdlib/plugin.rs
@@ -31,7 +31,7 @@ pub extern "C" fn kclvm_plugin_init(
 #[runtime_fn]
 pub unsafe extern "C" fn kclvm_plugin_invoke(
     ctx: *mut kclvm_context_t,
-    method: *const i8,
+    method: *const c_char,
     args: *const kclvm_value_ref_t,
     kwargs: *const kclvm_value_ref_t,
 ) -> *const kclvm_value_ref_t {

--- a/kclvm/runtime/src/value/api.rs
+++ b/kclvm/runtime/src/value/api.rs
@@ -193,7 +193,7 @@ pub unsafe extern "C" fn kclvm_value_Str(
 ) -> *mut kclvm_value_ref_t {
     let ctx = mut_ptr_as_ref(ctx);
     unsafe {
-        if v.is_null() || *v == '\0' as i8 {
+        if v.is_null() || *v == '\0' as c_char {
             return new_mut_ptr(ctx, ValueRef::str(""));
         }
     }
@@ -543,7 +543,7 @@ pub unsafe extern "C" fn kclvm_value_to_str_value(
 pub unsafe extern "C" fn kclvm_value_Str_ptr(p: *const kclvm_value_ref_t) -> *const kclvm_char_t {
     let p = ptr_as_ref(p);
     match &*p.rc.borrow() {
-        Value::str_value(ref v) => v.as_ptr() as *const i8,
+        Value::str_value(ref v) => v.as_ptr() as *const c_char,
         _ => std::ptr::null(),
     }
 }
@@ -629,7 +629,7 @@ pub unsafe extern "C" fn kclvm_value_function_invoke(
             // Normal kcl function, call directly
             } else if func.is_external {
                 let name = format!("{}\0", func.name);
-                kclvm_plugin_invoke(ctx, name.as_ptr() as *const i8, args, kwargs)
+                kclvm_plugin_invoke(ctx, name.as_ptr() as *const c_char, args, kwargs)
             } else {
                 args_ref.list_append_unpack_first(closure);
                 let args = args_ref.clone().into_raw(ctx_ref);

--- a/kclvm/src/lib.rs
+++ b/kclvm/src/lib.rs
@@ -16,7 +16,7 @@ use std::sync::Arc;
 ///
 /// args is a ExecProgramArgs JSON string.
 #[no_mangle]
-pub unsafe extern "C" fn kclvm_cli_run(args: *const i8, plugin_agent: *const i8) -> *const i8 {
+pub unsafe extern "C" fn kclvm_cli_run(args: *const c_char, plugin_agent: *const i8) -> *const i8 {
     let prev_hook = std::panic::take_hook();
 
     // disable print panic info
@@ -52,7 +52,7 @@ pub unsafe extern "C" fn kclvm_cli_run(args: *const i8, plugin_agent: *const i8)
 }
 
 /// KCL CLI run function CAPI.
-fn kclvm_cli_run_unsafe(args: *const i8, plugin_agent: *const i8) -> Result<String, String> {
+fn kclvm_cli_run_unsafe(args: *const c_char, plugin_agent: *const i8) -> Result<String, String> {
     let mut args =
         ExecProgramArgs::from_str(unsafe { std::ffi::CStr::from_ptr(args) }.to_str().unwrap());
     args.plugin_agent = plugin_agent as u64;

--- a/scripts/docker/kcl-builder-arm64/Dockerfile
+++ b/scripts/docker/kcl-builder-arm64/Dockerfile
@@ -1,0 +1,27 @@
+# Copyright The KCL Authors. All rights reserved.
+# Builder for kcllang/kcl-builder-arm64 image
+
+FROM arm64v8/ubuntu:20.04
+
+#RUN uname -a
+#RUN cat /etc/os-release
+
+RUN apt-get update
+
+RUN apt-get install -y curl make gcc git zlib-dev zlib1g-dev
+
+# rust
+RUN curl https://sh.rustup.rs -sSf | bash -s -- -y
+ENV CARGO_NET_GIT_FETCH_WITH_CLI=true
+
+RUN cargo version
+RUN rustc --version
+
+# clang12
+RUN apt-get install -y clang-12 lld-12
+RUN ln -sf /usr/bin/clang-12   /usr/bin/clang
+RUN ln -sf /usr/bin/wasm-ld-12 /usr/bin/wasm-ld
+
+WORKDIR /root
+
+CMD ["bash"]

